kaldb/src/main/java/com/slack/kaldb/clusterManager/SnapshotDeletionService.java

```diff
@@ -189,22 +190,36 @@ protected int deleteExpiredSnapshotsWithoutReplicas() {
                               // metadata and try again on the next run.
                               URI snapshotUri = URI.create(snapshotMetadata.snapshotPath);
                               LOG.info("Starting delete of snapshot {}", snapshotMetadata);
-                              if (s3BlobFs.exists(snapshotUri)) {
-                                // Ensure that the file exists before attempting to delete, in case
-                                // the previous run successfully deleted the object but failed the
-                                // metadata delete. Otherwise, this would be expected to perpetually
-                                // fail deleting a non-existing file.
-                                if (s3BlobFs.delete(snapshotUri, true)) {
-                                  snapshotMetadataStore.deleteSync(snapshotMetadata);
+
+                              // There can be scenarios where the indexer just gets killed by kube
+                              // (say a healthcheck fail)
+                              // In that case the indexer doesn't remove live chunks and the
+                              // snapshot remains in the /LIVE znode parent path
+                              // Since it's a live snapshot, we don't have a URL associated with it
+                              // and the pod is dead anyways
+                              if (SnapshotMetadata.isLive(snapshotMetadata)) {
+                                snapshotMetadataStore.deleteSync(snapshotMetadata);
+                              } else {
+                                if (s3BlobFs.exists(snapshotUri)) {
+                                  // Ensure that the file exists before attempting to delete, in
+                                  // case
+                                  // the previous run successfully deleted the object but failed the
+                                  // metadata delete. Otherwise, this would be expected to
+                                  // perpetually
+                                  // fail deleting a non-existing file.
+                                  if (s3BlobFs.delete(snapshotUri, true)) {
```

Comment:
So we have two options here on how to delete the S3 partial data for lives
1. We have a S3 bucket policy at the AWS level for 60 days currently. Just let S3 expire the data out
2. Reconstruct the s3 url - we know it's `s3://bucket-path/snapshotId`
3. Add  the snapshotPath for lives - the indexer could do this ( since it knows eventually where the path will be )



path of a live vs non live

```
[zk: localhost:2181(CONNECTED) 3] get /traces-prod-preprocessor/partitioned_snapshot/19584_9/log_1692231262_cc409877-dc06-4269-95c5-26a0c25b67da
{
  "name": "log_1692231262_cc409877-dc06-4269-95c5-26a0c25b67da",
  "snapshotId": "log_1692231262_cc409877-dc06-4269-95c5-26a0c25b67da",
  "snapshotPath": "s3://<bucket-name>/log_1692231262_cc409877-dc06-4269-95c5-26a0c25b67da/",
  "startTimeEpochMs": "1692090307335",
  "endTimeEpochMs": "1692223256109",
  "maxOffset": "36483141305",
  "partitionId": "180",
  "indexType": "LOGS_LUCENE9"
}```

non-live
```
[zk: localhost:2181(CONNECTED) 3] get /traces-prod-preprocessor/partitioned_snapshot/19584_9/log_1692231262_cc409877-dc06-4269-95c5-26a0c25b67da
{
  "name": "log_1692231262_cc409877-dc06-4269-95c5-26a0c25b67da",
  "snapshotId": "log_1692231262_cc409877-dc06-4269-95c5-26a0c25b67da",
  "snapshotPath": "s3://<bucket-name>/log_1692231262_cc409877-dc06-4269-95c5-26a0c25b67da/",
  "startTimeEpochMs": "1692090307335",
  "endTimeEpochMs": "1692223256109",
  "maxOffset": "36483141305",
  "partitionId": "180",
  "indexType": "LOGS_LUCENE9"
}```

Reply:
I was tracking down another set of errors and saw this error

```
com.slack.kaldb.metadata.core.InternalMetadataStoreException: Error deleting node under at path: LIVE_log_1692690142_3e9d9ee8-ec58-44f3-a764-4c1ed7d26d3a_10.197.67.221
	at com.slack.kaldb.metadata.core.KaldbMetadataStore.deleteSync(KaldbMetadataStore.java:153)
	at com.slack.kaldb.chunk.IndexingChunkImpl.preClose(IndexingChunkImpl.java:94)
	at com.slack.kaldb.chunk.ReadWriteChunk.close(ReadWriteChunk.java:178)
	at com.slack.kaldb.chunkManager.IndexingChunkManager.shutDown(IndexingChunkManager.java:366)

org.apache.zookeeper.KeeperException$ConnectionLossException: KeeperErrorCode = ConnectionLoss
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:102)
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:54)
	at org.apache.curator.x.async.details.BackgroundProcs.lambda$makeProc$2(BackgroundProcs.java:61)
	at org.apache.curator.x.async.details.InternalCallback.processResult(InternalCallback.java:53)
	at org.apache.curator.framework.imps.CuratorFrameworkImpl.sendToBackgroundCallback(CuratorFrameworkImpl.java:926)
	at org.apache.curator.framework.imps.CuratorFrameworkImpl.checkBackgroundRetry(CuratorFrameworkImpl.java:878)
```

We see a few of these every deploy. This means we don't cleanup the live metadata which then leads to the error initially linked to this Jira

Reply:
Some more notes

Today the cache node receives an assignment and it downloads the chunk via `BlobFsUtils#copyFromS3` which recreates the snapshotUrl. So the `snapshotPath` withing SnapshotMetada is never used.

Secondly,

We set the snapshotPath correct https://github.com/slackhq/kaldb/blob/master/kaldb/src/main/java/com/slack/kaldb/chunk/ReadWriteChunk.java#L247
but in https://github.com/slackhq/kaldb/blob/master/kaldb/src/main/java/com/slack/kaldb/chunk/IndexingChunkImpl.java#L63 , the live snapshot update keeps the `LIVE` snapshotPath which is the problem.